### ISSUE TITLE
Use TracePoint to get method information

### ISSUE THREAD

**Issue by muzfuz** [ruby-analysis/delfos#12]
Related to https://github.com/markburns/delfos/issues/9

Opening this issue to start a conversation around this.

It seems like the first step is to remove some of the metaprogramming and magic and take a lighter touch to the monkeypatching of BasicObject.  

Will post any related pull requests here.

**Comment by markburns** [comment]
Yeah we don't even need the monkey patching.

I'd probably be interested in seeing a performance comparison before making a decision.

**Comment by muzfuz** [comment]
So if I understand this correctly -- the idea would be to wrap _around_ method calls instead of dynamically overwriting methods by hooking into `method_added` and `singleton_method_added`?

 I think we'll still need to make use of `binding_of_caller` since it looks like that information isn't exposed to us by TracePoint / `set_trace_func`

**Comment by markburns** [comment]
Yeah that's right. Actually I'm hoping this isn't the reason we gave up on
TracePoint. That might be the case, that we don't have caller information.

However we could probably continue to get this by using he same pattern as
in ExecutionChain with thread local variables keeping track of the call
stack

**Comment by markburns** [comment]
I think my concerns were irrelevant. 
The main reasons against `TracePoint` were just performance and not having access to the arguments. But these are no longer relevant.
We can get everything we need like this:

```ruby
require "binding_of_caller"
thread = Thread.current
thread[:foo] = 42


class A
  def something
    t = Thread.new do
      t = Thread.current
      t[:foo] = 100

      B.new.another_thing
    end

    t.join
  end
end

class B
  def another_thing
    1
  end
end

trace = TracePoint.new(:call) do |tp|
  puts "-" * 80
  thread = Thread.current
  puts "foo: #{thread[:foo]}"

  if tp.defined_class == A || tp.defined_class == B
    puts "caller: #{binding.of_caller(2).receiver}"
    puts "  __FILE__: #{binding.of_caller(2).eval '__FILE__'}"
    puts "  __LINE__: #{binding.of_caller(2).eval '__LINE__'}"
    puts "  __method__: #{binding.of_caller(2).eval '__method__'}"
  end
end


trace.enable

A.new.something
```

```
--------------------------------------------------------------------------------
foo: 42
caller: main
  __FILE__: tp.rb
  __LINE__: 41
  __method__: nil
--------------------------------------------------------------------------------
foo: 100
caller: #<A:0x007fd60b019698>
  __FILE__: tp.rb
  __LINE__: 12
  __method__: :something
```

**Comment by muzfuz** [comment]
Right, I'll try this morning to put together a branch that starts this work.

Sounds like a pretty major re-write -- or at the very least, it'll lead to the deletion of a lot of classes and make this a much smaller gem.

**Comment by markburns** [comment]
Yeah exactly. It's more of a removal of a huge amount of metaprogramming workarounds used to replicate similar behaviour to TracePoint.
In fact, looking at it, the main deletion will be the https://github.com/markburns/delfos/tree/master/lib/delfos/patching folder.
So should be easy. And a testament to organising code around packaging concepts.

**Comment by muzfuz** [comment]
I suspect the `method_logging` folder can go as well.

Yeah, organising code around the concept of packages makes a lot of sense!

**Comment by markburns** [comment]
I think it might be easier to continue to use the `method_logging` stuff?
We still need to extract file/line_number etc from `binding.of_caller(OFFSET)`

**Comment by markburns** [comment]
Updated TracePoint usage example here: https://gist.github.com/markburns/632a50228d3bf9ba43d0ff62ce2adc33

**Comment by muzfuz** [comment]
https://github.com/markburns/delfos/pull/13

**Comment by markburns** [comment]
Fixed in #13 - @muzfuz - would be great if you get a chance to run this and find any issues. 
I'm assuming we'll find a bunch so after *those* are fixed it should be good for a new release.
